From 9e5b4c350e1bc8ac6712402c184e19d0e89ca3bd Mon Sep 17 00:00:00 2001
From: Nicolas Graves <ngraves@ngraves.fr>
Date: Wed, 11 Dec 2024 10:36:23 +0100
Subject: [PATCH] rde: sway: Allow keyboard-specific workspace keys

I use a french keyboard layout (azerty), which has the following
workspace keys:
'(ampersand eacute quotedbl apostrophe parenleft minus egrave
underscore ccedilla agrave)
This should probably be managed by reading xkb files at config time,
but until we can do that properly, this is a temporary solution.
---
 src/rde/features/wm.scm | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/src/rde/features/wm.scm b/src/rde/features/wm.scm
index 5531badc..7d088dc9 100644
--- a/src/rde/features/wm.scm
+++ b/src/rde/features/wm.scm
@@ -120,7 +120,8 @@
           ;; Logo key. Use Mod1 for Alt.
           (sway-mod 'Mod4)
           (add-keyboard-layout-to-config? #t)
-          (xwayland? #f))
+          (xwayland? #f)
+          (workspace-keys (append (iota 9 1) '(0))))
   "Setup and configure sway."
   (ensure-pred sway-config? extra-config)
   (ensure-pred boolean? add-keyboard-layout-to-config?)
@@ -133,6 +134,7 @@
   (ensure-pred file-like? xdg-desktop-portal)
   (ensure-pred file-like? xdg-desktop-portal-gtk)
   (ensure-pred file-like? xdg-desktop-portal-wlr)
+  (ensure-pred list? workspace-keys)
 
   (define (sway-home-services config)
     "Returns home services related to sway."
@@ -245,12 +247,12 @@
             (,#~"\n\n# Moving around workspaces:")
             (bindsym $mod+tab workspace back_and_forth)
             ,@(append-map
-               (lambda (x)
-                 `((bindsym ,(format #f "$mod+~a" (modulo x 10))
-                            workspace number ,x)
-                   (bindsym ,(format #f "$mod+Shift+~a" (modulo x 10))
-                            move container to workspace number ,x)))
-               (iota 10 1))
+               (lambda (key number)
+                 `((bindsym ,(format #f "$mod+~a" key)
+                            workspace number ,number)
+                   (bindsym ,(format #f "$mod+Shift+~a" key)
+                            move container to workspace number ,number)))
+               workspace-keys (iota 10 1))
 
             (,#~"\n\n# Scratchpad settings:")
             (bindsym --to-code $mod+Shift+minus move scratchpad)
-- 
2.46.0

